### PR TITLE
Update the weight function for Mark Cache and Min Max Index Cache

### PR DESCRIPTION
**Pull request by hongyunyan**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tiflash/issues/8050

Problem Summary:
Update the weight function for Mark Cache and Min Max Index Cache

### What is changed and how it works?
We only calculate the inner value weight as the cost of LRUCache now, while sometimes the memory cost of key and structural part also occupy a lot, especially when the pack number is quite small.

Thus, we update the weight function of mark cache and minmax index cache, to provider more accurate memory calculation(while it's still an approximate memory usage, but more accurate approximat).

Besides, we add the metrics for minmax cache index, and add their correponding panel in grafana.

In test we can see the whole memory gap is similar with the memory gap of mark cache and minmax index cache.

image

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/DeltaMerge/Index/MinMaxIndex.h [hunk 1]
```diff
@@ -60,7 +60,8 @@ class MinMaxIndex
 
     size_t byteSize() const
     {
-        return sizeof(UInt8) * has_null_marks->size() + sizeof(UInt8) * has_value_marks->size() + minmaxes->byteSize();
+        return sizeof(UInt8) * has_null_marks->size() + sizeof(UInt8) * has_value_marks->size() + minmaxes->byteSize()
+            + 3 * sizeof(PaddedPODArray<UInt8>);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/Index/MinMaxIndex.h:64]: Add some comments about where these values comes from

File: dbms/src/Storages/DeltaMerge/Remote/RNDeltaIndexCache.h [hunk 2]
```diff
@@ -85,7 +85,7 @@ class RNDeltaIndexCache : private boost::noncopyable
 
     struct CacheValueWeight
     {
-        size_t operator()(const CacheValue & v) const { return v.bytes; }
+        size_t operator()(const CacheKey & /*key*/, const CacheValue & v) const { return v.bytes; }
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/Remote/RNDeltaIndexCache.h:88]: We'd better add `sizeof(CacheKey)` here cause the key is composed by plenty of members

File: dbms/src/Interpreters/AsynchronousMetrics.cpp [hunk 3]
```diff
@@ -196,6 +196,14 @@ void AsynchronousMetrics::update()
...
     }
 
+    {
+        if (auto min_max_cache = context.getMinMaxIndexCache())
+        {
+            set("MinMaxIndexCacheBytes", min_max_cache->weight());
+            set("MinMaxIndexFiles", min_max_cache->count());
+        }
+    }
+
```
Comment (JaySon-Huang) [inline dbms/src/Interpreters/AsynchronousMetrics.cpp:206]: We'd better add metrics about `RNDeltaIndexCache` when it is not empty

File: dbms/src/Storages/DeltaMerge/Index/MinMaxIndex.h [hunk 4]
```diff
@@ -116,7 +117,24 @@ class MinMaxIndex
...
+        auto str_len = key.size(); // key_len
+        auto key_memory_usage = sizeof(String); // String struct memory cost
+
+        // 3. the memory cost of hash table
+        auto unordered_map_memory_usage = 28; // hash table struct approximate memory cost
+
+        // 4. the memory cost of LRUQueue
+        auto list_memory_usage = sizeof(std::list<String>); // list struct memory cost
+
+        return index_memory_usage + cells_memory_usage + str_len * 2 + key_memory_usage * 2 + unordered_map_memory_usage
```
[the remark this answers is not in the document]
Comment (hongyunyan) [reply dbms/src/Storages/DeltaMerge/Index/MinMaxIndex.h:138]: One for LRUCache::Queue and one for key of LRUCache::Cells

**Review by JaySon-Huang — APPROVED** [review]

**Pull request merged**